openwisp_network_topology/integrations/device/base/models.py

```diff
@@ -185,3 +215,264 @@ def trigger_device_updates(cls, link):
             if 'openwisp_monitoring.device' in settings.INSTALLED_APPS:
                 run_checks = import_string(trigger_device_checks_path)
                 run_checks.delay(device_node.device.pk, recovery=link.status == 'up')
+
+
+class AbstractWifiMesh(UUIDModel):
+    _NODE_PROPERTIES = [
+        'ht',
+        'vht',
+        'he',
+        'mfp',
+        'wmm',
+        'vendor',
+    ]
+    _LINK_PROPERTIES = [
+        'auth',
+        'authorized',
+        'noise',
+        'signal',
+        'signal_avg',
+        'mesh_llid',
+        'mesh_plid',
+        'mesh_plink',
+        'mesh_non_peer_ps',
+    ]
+
+    topology = models.ForeignKey(
+        get_model_name('topology', 'Topology'), on_delete=models.CASCADE
+    )
+    mesh_id = models.CharField(
+        max_length=32, null=False, blank=False, verbose_name=_('Mesh ID')
+    )
+
+    class Meta:
+        abstract = True
+
+    @classmethod
+    def create_topology(cls, organization_ids, discard_older_data_time):
+        Link = load_model('topology', 'Link')
+        for org_id in organization_ids:
+            intermediate_topologies = cls._create_intermediate_topologies(
+                org_id, discard_older_data_time
+            )
+            if not intermediate_topologies:
+                Link.objects.filter(
+                    topology__wifimesh__isnull=False, organization_id=org_id
+                ).exclude(topology__wifimesh__in=intermediate_topologies.keys()).update(
+                    status='down'
+                )
+                continue
+            cls._create_topology(intermediate_topologies, org_id)
+
+    @classmethod
+    def _create_intermediate_topologies(cls, organization_id, discard_older_data_time):
+        DeviceData = load_model('device_monitoring', 'DeviceData')
+        intermediate_topologies = {}
+        query = DeviceData.objects.filter(organization_id=organization_id).only(
+            'mac_address'
+        )
+        discard_older_data_time = now() - timedelta(seconds=discard_older_data_time)
+        for device_data in query.iterator():
+            try:
+                assert device_data.data is not None
+                data_timestamp = datetime.fromisoformat(device_data.data_timestamp)
+                assert data_timestamp > discard_older_data_time
+            except (AttributeError, AssertionError):
+                continue
+            mesh_interfaces = AbstractWifiMesh._get_mesh_interfaces(device_data)
```

Comment:
I think looping over the interfaces twice is not a great approach which we shouldn't adopt as a standard.
Here the code loops over all the interfaces first, then loops again on the mesh interfaces. It seems to me that the list resutling from this method is not used anywhere else.
This operation is slightly inefficient and we should avoid it if possible, I think you could simply loop over the interfaces as in the method.

If you need to simplify the code, then separate the code block in the loop to another method. Example below.